lib/sandbox.js

```diff
@@ -0,0 +1,50 @@
+'use strict';
+
+const vm = require('vm');
+
+const VM_OPTIONS = {
+  filename: 'iuam-challenge.js',
+  contextOrigin: 'cloudflare:iuam-challenge.js',
+  contextCodeGeneration: { strings: true, wasm: false },
+  timeout: 5000
+};
+
+module.exports = { eval: evaluate, Context };
+
+function evaluate (code, ctx) {
+  return vm.runInNewContext(code, ctx, VM_OPTIONS);
+}
+
+// Global context used to evaluate standard IUAM JS challenge
+function Context (options) {
+  if (!options) options = { body: '', hostname: '' };
+
+  const body = options.body;
+  const href = 'http://' + options.hostname + '/';
+  const cache = Object.create(null);
+  const keys = [];
+
+  this.atob = function (str) {
+    return Buffer.from(str, 'base64').toString('binary');
+  };
+
+  // Used for eval during onRedirectChallenge
+  this.location = { reload: function () {} };
+
+  this.document = {
+    createElement: function () {
+      return { firstChild: { href: href } };
+    },
+    getElementById: function (id) {
+      if (keys.indexOf(id) === -1) {
+        const re = new RegExp(' id=[\'"]?' + id + '[^>]*>([^<]*)');
```

Comment:
@codemanki Note that this does not account for self closing nodes. Should the challenge ever use `getElementByid` to access a self closing node it won't fail but it will return the incorrect `innerHTML`. It's easy enough to fix if that ever becomes a problem.

Reply:
Also not resolved, this is the same as on the master branch. (published)